lib/ABI/ModelAlignment.cpp

```diff
@@ -0,0 +1,171 @@
+/// \file ModelAlignment.cpp
+/// \brief
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "revng/ABI/ModelAlignment.h"
+#include "revng/ABI/RuntimeTraitAccessors.h"
+#include "revng/Model/Binary.h"
+
+namespace abi {
+
+static RecursiveCoroutine<std::optional<uint64_t>>
+naturalAlignmentImpl(model::VerifyHelper &VH,
+                     const model::Type &Type,
+                     model::ABI::Values ABI);
+static RecursiveCoroutine<std::optional<uint64_t>>
+naturalAlignmentImpl(model::VerifyHelper &VH,
+                     const model::QualifiedType &Type,
+                     model::ABI::Values ABI);
+
+template<typename FieldType>
+static RecursiveCoroutine<std::optional<uint64_t>>
+aggregateAlignmentImpl(model::VerifyHelper &VH,
+                       const SortedVector<FieldType> Fields,
+                       model::ABI::Values ABI) {
+  uint64_t Alignment = 0;
+
+  for (const auto &Field : Fields) {
+    if (auto A = rc_recur naturalAlignmentImpl(VH, Field.Type, ABI))
+      Alignment = std::max(Alignment, *A);
+    else
+      rc_return std::nullopt;
+  }
+
+  rc_return Alignment;
+}
+
+static RecursiveCoroutine<std::optional<uint64_t>>
+naturalAlignmentImpl(model::VerifyHelper &VH,
+                     const model::Type &Type,
+                     model::ABI::Values ABI) {
+  auto MaybeAlignment = VH.size(&Type);
+  if (MaybeAlignment)
+    rc_return MaybeAlignment;
+
+  // This code assumes that the type `Type` is well formed.
+  uint64_t Alignment;
+
+  switch (Type.Kind) {
+  case model::TypeKind::RawFunctionType:
+  case model::TypeKind::CABIFunctionType:
+    // Function prototypes have no size - hence no alignment.
+    Alignment = 0;
+    break;
+
+  case model::TypeKind::PrimitiveType:
+    // The alignment of primitives is simple to figure out based on the abi
+    if (const auto *P = llvm::cast<model::PrimitiveType>(&Type);
```

Comment:
Hoist `P` out of the `if`.

Reply:
Keeping it inside let's us skip one less set of curly braces though 😅.

Not sure why, but I dislike having to write
```cpp
case SomeValue: {
  // . . . 
}
```

Having the variable defined inside `if` affixes its lifetime to the `if` braces removing the need to explicitly manage it.

Reply:
Having a `llvm::cast` in an `if` is misleading, usually you use `dyn_cast`.
About the graphs, I don't think the battle against those is viable/useful `:P`

Reply:
1. `llvm::cast` is what `size()` used, so I just followed it.
2. But it's one less line (no need for the closing brace), don't you like those negative-sum commits? `:P`

Reply:
We're overdiscussing this